transport/grpc/handler.go

```diff
@@ -98,6 +114,35 @@ func getServiceAndProcedureFromMethod(streamMethod string) (string, string, erro
 	return service, procedure, nil
 }
 
+func getContextMetadata(ctx context.Context) (metadata.MD, error) {
+	contextMetadata, ok := metadata.FromContext(ctx)
+	if !ok || contextMetadata == nil {
+		return nil, errors.New("Could not extract metadata information from context")
+	}
+	return contextMetadata, nil
+}
+
+func getCaller(ctxMetadata metadata.MD) (string, error) {
+	// TODO Make a gRPC outbounds require a header for caller and enforce it at inbounds
+	return extractMetadataHeader(ctxMetadata, CallerHeader)
+}
+
+func getEncoding(ctxMetadata metadata.MD) (string, error) {
+	// TODO Add a pull request to gRPC to add encoding via content-type to contexts so we don't have to set it manually ourselves
+	return extractMetadataHeader(ctxMetadata, EncodingHeader)
+}
+
+func extractMetadataHeader(ctxMetadata metadata.MD, header string) (string, error) {
+	headerList, ok := ctxMetadata[header]
+	if !ok {
+		return "", fmt.Errorf("Couldn't extract header:(%s) from Context Metadata (%v)", header, ctxMetadata)
```

Comment:
Also, it's not that you couldn't extract the headers. They weren't there. This and the next error can probably be consolidated. In fact, I don't know if you need to check for `ok` at all. If it's `!ok`, you'll get `nil` which is a valid slice with `len(headerList) == 0`.